src/Azure/Orleans.Transactions.AzureStorage/TransactionalState/AzureTableTransactionalStateStorage.cs

```diff
@@ -0,0 +1,184 @@
+                if (string.IsNullOrEmpty(this.key.ETag))
+                    batchOperation.Insert(this.key);
+                else
+                    batchOperation.Replace(this.key);
+
+                // add new states
+                List<Tuple<string,long>> stored = this.states.Select(s => Tuple.Create(s.TransactionId, s.SequenceId)).ToList();
+                List<StateEntity> newStates = new List<StateEntity>();
+                foreach (PendingTransactionState<TState> pendingState in statesToPrepare.Where(p => !stored.Contains(Tuple.Create(p.TransactionId, p.SequenceId))))
+                {
+                    var newState = StateEntity.Create(this.JsonSettings, this.partition, pendingState);
+                    newStates.Add(newState);
+                    batchOperation.Insert(newState);
+                }
+
+                if (batchOperation.Count > AzureTableConstants.MaxBatchSize)
+                {
+                    this.logger.LogError("Too many pending states. PendingStateCount {PendingStateCount}.", batchOperation.Count);
+                    throw new InvalidOperationException($"Too many pending states. PendingStateCount {batchOperation.Count}");
+                }
+
+                await table.ExecuteBatchAsync(batchOperation).ConfigureAwait(false);
+                this.states.AddRange(newStates);
+                return this.key.ETag;
+            }
+            catch (Exception ex)
+            {
+                this.logger.LogError("Transactional state persist failed {Exception}.", ex);
+                throw;
+            }
+        }
+
+        public async Task<string> Confirm(string expectedETag, string metadata, string transactionIdToCommit)
+        {
+            try
+            {
+                // only update storage if transaction id is greater then previously committed
+                if (string.Compare(transactionIdToCommit, this.key.CommittedTransactionId) <= 0)
+                    return this.key.ETag;
+
+                if (!TryFindState(transactionIdToCommit, out TState state))
+                {
+                    this.logger.LogCritical("Transactional state non-recoverable error.  Attempting to confirm a transaction {TransactionId} for which no state exists.", transactionIdToCommit);
+                    throw new InvalidOperationException($"Transactional state non-recoverable error.  Attempting to confirm a transaction {transactionIdToCommit} for which no state exists.");
+                }
+
+                this.key.ETag = expectedETag;
+                this.key.Metadata = metadata;
+                this.key.CommittedTransactionId = transactionIdToCommit;
+                await WriteKey();
+                var dead = this.states.Where(p => string.Compare(p.TransactionId, transactionIdToCommit) <0).ToList();
+                this.states = this.states.Where(p => string.Compare(p.TransactionId, transactionIdToCommit) >= 0).ToList();
+                Cleanup(dead).Ignore();
```

Comment:
Why is the cleanup asynchronous, separate from the commit? Is that really necessary? I think doing everything in a single batch is cleaner and peforms better.

Reply:
Doing it all in a single operation would be cleaner.  Agreed
Doing it all in a single operation would be faster.  Agreed

I elected to perform cleanup as a separate operation to reduce cost and risk of the critical operation.  The cleanup is optional in that it does not affect the core behavior other than making loads slower over time.  I didn't want to risk storage errors in the confirm path, which is critical.

There are other patterns that could be used:
* We could leave the entities in storage and scope the load query by starting it with the committed transaction id, but that means we couldn't perform the key and state queries in parallel.
* We could setup some sort of background operation that periodically cleaned up the table, but then we'd need to manage that operation.
* We could clean up the table once on load, but that would still effect load time.

I'm open to other solutions, I just don't want to introduce unnecessary risk to the commit operation.

Reply:
> reduce cost and risk of the critical operation

I am not sure I see a difference in the "criticality" ... by the time this is called the transaction has already irreversably committed (i.e. commit record is present in TM log). So, everything that is happening here is really a "cleanup" operation.

Reply:
> I am not sure I see a difference in the "criticality"
Deleting older states is optional cleanup.
Confirm success is crtical to operation.  This operation must eventually succeed and failures negativly effect the performacne.

Authoritatively you are correct, this is just notifying the grain that the transaction has been commited. but this operation *must* eventually succeed, and every failure keeps following transaction states in a pending state and dependent on the committed transaction.  This does not lead to errors, assuming this operation eventually succeeds, but does negatively affect the performance of the system.

This operation currently writes one single entity, which is as low risk as it gets, as opposed to writing a batch and dealing with the complexity (batch entity counts and size limits) and the risk in which that entails.  To risk this to perform a mostly optional cleanup operation in a cleaner and faster manner does not seem a good trade.

We can always cleanup and optimize the cleanup operation, but it is logically separable from the confirm operation which is not optional.